src/main/java/frc/robot/subsystems/ShooterSubsystem.java

```diff
@@ -0,0 +1,152 @@
+package frc.robot.subsystems;
+
+import com.revrobotics.CANSparkBase;
+import com.revrobotics.CANSparkLowLevel.MotorType;
+import com.revrobotics.CANSparkMax;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+
+public class ShooterSubsystem extends SubsystemBase {
+
+  private final CANSparkMax shooterTopMotor;
+  private final CANSparkMax shooterBottomMotor;
+
+  public ShooterSubsystem() {
+    shooterTopMotor =
+        new CANSparkMax(Constants.ShooterConstants.SHOOTER_MOTOR_TOP_ID, MotorType.kBrushless);
+    shooterBottomMotor =
+        new CANSparkMax(Constants.ShooterConstants.SHOOTER_MOTOR_BOTTOM_ID, MotorType.kBrushless);
+
+    shooterTopMotor.restoreFactoryDefaults();
+    shooterTopMotor.setSmartCurrentLimit(40);
+    shooterBottomMotor.restoreFactoryDefaults();
+    shooterBottomMotor.setSmartCurrentLimit(40);
+
+    shooterTopMotor.getPIDController().setP(Constants.ShooterConstants.SHOOTER_P);
+    shooterTopMotor.getPIDController().setI(Constants.ShooterConstants.SHOOTER_I);
+    shooterTopMotor.getPIDController().setD(Constants.ShooterConstants.SHOOTER_D);
+    shooterTopMotor.getPIDController().setFF(Constants.ShooterConstants.SHOOTER_FF);
+    shooterBottomMotor.getPIDController().setP(Constants.ShooterConstants.SHOOTER_P);
+    shooterBottomMotor.getPIDController().setI(Constants.ShooterConstants.SHOOTER_I);
+    shooterBottomMotor.getPIDController().setD(Constants.ShooterConstants.SHOOTER_D);
+    shooterBottomMotor.getPIDController().setFF(Constants.ShooterConstants.SHOOTER_FF);
+  }
+
+  // Fires the shooter.
+  public Command shooterFire(double distance) {
+    return startEnd(
+        () -> {
+          if (isShooterReady(distance)) {
+            setShooterSpeeds(calculateShooterSpeeds(distance));
+          }
+        },
+        () -> {});
+  }
+
+  // Idle shooter command; for default command purposes
+  public Command shooterIdle() {
+    return run(
+        () -> {
+          setShooterSpeeds(Constants.ShooterConstants.SHOOTER_IDLE_SPEEDS);
+        });
+  }
+
+  // Checks if shooter is ready.
+  public boolean isShooterReady(double distance) {
+    boolean shooterReady = false;
+    double minSpeedTolerance =
+        calculateShooterSpeeds(distance)[0]
+            * (1 - Constants.ShooterConstants.SHOOTER_SPEED_TOLERANCE);
+    double maxSpeedTolerance =
+        calculateShooterSpeeds(distance)[0]
+            * (1 + Constants.ShooterConstants.SHOOTER_SPEED_TOLERANCE);
+
+    if (minSpeedTolerance < shooterTopMotor.getEncoder().getVelocity()
+        & shooterTopMotor.getEncoder().getVelocity() < maxSpeedTolerance) {
+      shooterReady = true;
+    }
+
+    return shooterReady;
+  }
+
+  // Speed in RPM. Top is index 0, bottom is index 1.
+  public void setShooterSpeeds(double[] speeds) {
```

Comment:
I dislike representing non-sequential data in an array like this, because it is unnecessarily confusing. Instead, I think it would be better to either accept separate values for the upper vs lower motor, or accept a custom class which contains both. Given that this is a simple method which accepts only two parameters, it probably makes more sense to pass them in separately.

Reply:
However given that the motor speeds need to be tied together, it might make sense to use a custom class or an immutable.

Reply:
The issue is that it is impossible to _return_ multiple values, which is necessary in calculateShooterSpeeds(). However, I will use a custom class as recommended by Adam.

Reply:
Acknowledged in 5def812 and a1e8952, using Pair<Double, Double>